sky/serve/autoscalers.py

```diff
@@ -238,3 +264,168 @@ def _get_replica_ids_to_scale_down(num_limit: int) -> List[int]:
         if not scaling_options:
             logger.info('No scaling needed.')
         return scaling_options
+
+
+class SpotRequestRateAutoscaler(RequestRateAutoscaler):
+    """SpotRequestRateAutoscaler: Use spot to autoscale based on request rate.
+
+    This autoscaler uses spot instances to save cost.
+    """
+
+    NAME: Optional[str] = 'SpotRequestRateAutoscaler'
+
+    def __init__(self, spec: 'service_spec.SkyServiceSpec') -> None:
+        super().__init__(spec)
+        self.spot_placer: 'spot_policy.SpotPlacer' = (
+            spot_policy.SpotPlacer.from_spec(spec))
+
+    def handle_active_history(self, history: List[str]) -> None:
+        for zone in history:
+            self.spot_placer.handle_active(zone)
+
+    def handle_preemption_history(self, history: List[str]) -> None:
+        for zone in history:
+            self.spot_placer.handle_preemption(zone)
+
+    def _get_spot_resources_override_dict(self) -> Dict[str, Any]:
+        return {'use_spot': True, 'spot_recovery': None}
+
+    def _scale_spot_instances(
+        self, num_alive_spot: int,
+        launched_replica_infos: List['replica_managers.ReplicaInfo']
+    ) -> List[AutoscalerDecision]:
+
+        scaling_options = []
+        all_replica_ids_to_scale_down: List[int] = []
+        num_to_provision = (self.target_num_replicas + self.num_overprovision)
+
+        # Scale spot instances.
+        current_considered_zones: List[str] = []
+        if num_alive_spot < num_to_provision:
+            # Not enough spot instances, scale up.
+            num_spot_to_scale_up = num_to_provision - num_alive_spot
+            for _ in range(num_spot_to_scale_up):
+                spot_override = self._get_spot_resources_override_dict()
+                zone = self.spot_placer.select(launched_replica_infos,
+                                               current_considered_zones)
+                current_considered_zones.append(zone)
+                spot_override.update({'zone': zone})
+                logger.info(f'Chosen zone {zone} with {self.spot_placer}')
+                scaling_options.append(
+                    AutoscalerDecision(AutoscalerDecisionOperator.SCALE_UP,
+                                       target=spot_override))
+        elif num_alive_spot > num_to_provision:
+            # Too many spot instances, scale down.
+            num_spot_to_scale_down = num_alive_spot - num_to_provision
+            all_replica_ids_to_scale_down.extend(
+                RequestRateAutoscaler.get_replica_ids_to_scale_down(
+                    num_spot_to_scale_down,
+                    [info for info in launched_replica_infos if info.is_spot]))
+
+        for replica_id in all_replica_ids_to_scale_down:
+            scaling_options.append(
+                AutoscalerDecision(AutoscalerDecisionOperator.SCALE_DOWN,
+                                   target=replica_id))
```

Comment:
nit: Could we directly append scaling options, instead of store all id in a list and then adding it?